File: models/Submission.js
const { getDb } = require("../config/firebase");
const { FieldPath, Timestamp } = require("firebase-admin/firestore");

const COLLECTION = "submissions";
const SESSIONS_COLLECTION = "sessions";
const AGGREGATES_DOC = () => getDb().collection("system").doc("aggregates");

/**
 * @typedef {Object} Submission
 * @property {string}  id
 * @property {string}  fullName
 * @property {string}  normalizedName
 * @property {number}  score
 * @property {number}  total
 * @property {number}  percentage
 * @property {string}  prize
 * @property {string}  status   - "pending" | "completed"
 * @property {Date}    submittedAt
 * @property {Date}    [finalizedAt]
 * @property {number[]} [answers]
 */

const SubmissionModel = {
  /**
   * @returns {Promise<Submission[]>}
   */
  async findAll() {
    const snap = await getDb()
      .collection(COLLECTION)
      .orderBy("submittedAt", "desc")
      .get();

    return snap.docs.map((d) => ({ id: d.id, ...d.data() }));
  },

  /**
   * @param {string} id
   * @returns {Promise<Submission|null>}
   */
  async findById(id) {
    const doc = await getDb().collection(COLLECTION).doc(id).get();
    return doc.exists ? { id: doc.id, ...doc.data() } : null;
  },

  /**
   * Batch existence check for submission doc ids (admin list enrichment).
   * @param {string[]} ids
   * @returns {Promise<Set<string>>}
   */
  async idsThatExist(ids) {
    if (!ids.length) return new Set();
    const db = getDb();
    const refs = ids.map((id) => db.collection(COLLECTION).doc(id));
    const snaps = await db.getAll(...refs);
    return new Set(snaps.filter((s) => s.exists).map((s) => s.id));
  },

  /**
   * Paginated submissions (newest first).
   * @param {{ limit?: number, cursor?: { submittedAt: number, id: string }|null }} opts
   */
  async findPage({ limit = 50, cursor = null }) {
    const db = getDb();
    const cap = Math.min(Math.max(Number(limit) || 50, 1), 200);
    let q = db
      .collection(COLLECTION)
      .orderBy("submittedAt", "desc")
      .orderBy(FieldPath.documentId(), "desc")
      .limit(cap);

    if (cursor && typeof cursor.submittedAt === "number" && cursor.id) {
      const ts = Timestamp.fromMillis(cursor.submittedAt);
      q = q.startAfter(ts, cursor.id);
    }

    const snap = await q.get();
    const items = snap.docs.map((d) => ({ id: d.id, ...d.data() }));
    const last = snap.docs[snap.docs.length - 1];
    const hasMore = snap.docs.length === cap;
    const nextCursor =
      last && hasMore
        ? { submittedAt: last.get("submittedAt")?.toMillis?.() ?? 0, id: last.id }
        : null;

    return { items, nextCursor, hasMore };
  },

  /**
   * Prize award counts from `system/aggregates` (O(1)); rebuild from scan if missing.
   * @returns {Promise<Record<string, number>>}
   */
  async getPrizeCounts() {
    const doc = await AGGREGATES_DOC().get();
    const counts = doc.exists ? doc.data().prizeAwardCounts : null;
    if (counts && typeof counts === "object" && Object.keys(counts).length > 0) {
      return counts;
    }
    return this.rebuildPrizeAwardCounts();
  },

  /**
   * Full scan — use after deploy or if aggregates drift (admin maintenance).
   * @returns {Promise<Record<string, number>>}
   */
  async rebuildPrizeAwardCounts() {
    const db = getDb();
    const snap = await db.collection(COLLECTION).get();
    const counts = {};

    snap.forEach((d) => {
      const { prize } = d.data();
      if (prize && prize !== "pending" && prize !== "Nothing") {
        counts[prize] = (counts[prize] || 0) + 1;
      }
    });

    await AGGREGATES_DOC().set(
      { prizeAwardCounts: counts, rebuiltAt: new Date() },
      { merge: true }
    );

    return counts;
  },

  /**
   * Validate answers against Firestore questions, compute score server-side,
   * ignore any client-supplied prize/score. Uses a transaction for idempotency.
   *
   * @param {{ sessionId: string, fullName: string, normalizedName: string, answers: number[], userAgent?: string, ip?: string }} data
   * @returns {Promise<Submission>}
   */
  async create(data) {
    const QuestionModel = require("./Question");
    const db = getDb();
    const { sessionId, fullName, normalizedName, answers, userAgent, ip } = data;

    const questions = await QuestionModel.findAll();
    if (!questions.length) {
      const err = new Error("No questions configured.");
      err.code = "NO_QUESTIONS";
      throw err;
    }

    if (!Array.isArray(answers) || answers.length !== questions.length) {
      const err = new Error("answers must match the number of questions.");
      err.code = "INVALID_ANSWERS_LENGTH";
      throw err;
    }

    let score = 0;
    for (let i = 0; i < questions.length; i++) {
      const q = questions[i];
      const ans = answers[i];
      if (!Number.isInteger(ans) || ans < 0 || ans >= q.options.length) {
        const err = new Error(`Invalid answer index for question ${i}.`);
        err.code = "INVALID_ANSWER_INDEX";
        throw err;
      }
      if (ans === q.correctIndex) score += 1;
    }

    const total = questions.length;
    if (score > total) {
      const err = new Error("Invalid score.");
      err.code = "INVALID_SCORE";
      throw err;
    }

    const percentage = Math.round((score / total) * 100);
    const prize = score === total ? "pending" : "Nothing";

    const submissionRef = db.collection(COLLECTION).doc(sessionId);
    const sessionRef = db.collection(SESSIONS_COLLECTION).doc(sessionId);

    return db.runTransaction(async (t) => {
      const [subSnap, sessionSnap] = await Promise.all([
        t.get(submissionRef),
        t.get(sessionRef),
      ]);

      if (!sessionSnap.exists) {
        const err = new Error("No registration found for this session.");
        err.code = "NO_SESSION";
        throw err;
      }

      if (subSnap.exists) {
        return { id: subSnap.id, ...subSnap.data() };
      }

      const now = new Date();
      const payload = {
        sessionId,
        fullName: fullName.toUpperCase(),
        normalizedName,
        score,
        total,
        percentage,
        prize,
        answers,
        userAgent: userAgent || "unknown",
        ip: typeof ip === "string" ? ip : "unknown",
        status: prize === "pending" ? "pending" : "completed",
        submittedAt: now,
      };

      t.set(submissionRef, payload);
      t.set(
        sessionRef,
        {
          hasPlayed: true,
          playedAt: now,
          score,
          percentage,
          prize,
          status: prize === "pending" ? "pending" : "completed",
        },
        { merge: true }
      );

      return { id: sessionId, ...payload };
    });
  },

  /**
   * Atomically assign wheel prize when submission is still "pending".
   * Increments `system/aggregates.prizeAwardCounts` when `isRealPrize` is true.
   * @param {string} sessionId
   * @param {string} prizeName
   * @param {boolean} isRealPrize
   * @returns {Promise<{ finalized: boolean, previousPrize?: string }>}
   */
  async finalizeSpinPrize(sessionId, prizeName, isRealPrize) {
    const db = getDb();
    const now = new Date();
    const submissionRef = db.collection(COLLECTION).doc(sessionId);
    const sessionRef = db.collection(SESSIONS_COLLECTION).doc(sessionId);
    const statsRef = AGGREGATES_DOC();

    return db.runTransaction(async (t) => {
      const subSnap = await t.get(submissionRef);
      if (!subSnap.exists) {
        const err = new Error("Submission not found.");
        err.code = "NOT_FOUND";
        throw err;
      }

      const d = subSnap.data();
      if (d.prize && d.prize !== "pending") {
        return { finalized: false, previousPrize: d.prize };
      }

      t.update(submissionRef, {
        prize: prizeName,
        status: "completed",
        finalizedAt: now,
      });
      t.set(
        sessionRef,
        {
          prize: prizeName,
          status: "completed",
        },
        { merge: true }
      );

      if (isRealPrize) {
        const statsSnap = await t.get(statsRef);
        const prev = statsSnap.exists ? statsSnap.data().prizeAwardCounts || {} : {};
        const next = { ...prev, [prizeName]: (prev[prizeName] || 0) + 1 };
        t.set(statsRef, { prizeAwardCounts: next, updatedAt: now }, { merge: true });
      }

      return { finalized: true };
    });
  },

  /**
   * @param {string} id
   * @returns {Promise<void>}
   */
  async delete(id) {
    const db = getDb();
    const sub = await this.findById(id);
    const batch = db.batch();

    batch.delete(db.collection(COLLECTION).doc(id));

    if (sub) {
      const regSnap = await db.collection("registrations").doc(id).get();
      if (regSnap.exists) {
        const regData = regSnap.data();
        batch.delete(regSnap.ref);

        if (regData?.normalizedName) {
          batch.delete(db.collection("registrations").doc(`name_${regData.normalizedName}`));
        }

        if (regData?.ip && regData.ip !== "unknown") {
          const ipSnap = await db.collection("registrations").where("ip", "==", regData.ip).get();

          ipSnap.forEach((d) => {
            if (d.id !== id) {
              batch.delete(d.ref);
              batch.delete(db.collection(COLLECTION).doc(d.id));
              if (d.data()?.normalizedName) {
                batch.delete(db.collection("registrations").doc(`name_${d.data().normalizedName}`));
              }
            }
          });
        }
      }
    }

    await batch.commit();
  },
};

module.exports = SubmissionModel;

File: src/models/submission.rs
use crate::app_state::AppState;
use crate::error::{ApiError, ApiResult};
use crate::features::campaigns::infrastructure::CampaignPaths;
use crate::features::inventory::infrastructure::InventoryRepository;
use crate::models::prize::PrizeModel;
use crate::models::question::QuestionModel;
use crate::models::registration::RegistrationModel;
use crate::utils::firestore::millis_now;
use anyhow::anyhow;
use firestore::errors::{
    BackoffError, FirestoreError, FirestoreInvalidParametersError,
    FirestoreInvalidParametersPublicDetails,
};
use firestore::FirestoreQueryDirection;
use serde_json::{json, Map, Value};
use std::collections::HashSet;
use std::future::Future;
use std::pin::Pin;

const SUBMISSIONS_SUBCOL: &str = "submissions";
const SESSIONS_SUBCOL: &str = "sessions";
const REGISTRATIONS_SUBCOL: &str = "registrations";

pub struct SubmissionModel;

#[derive(Debug)]
pub struct SubmissionCreateInput {
    pub session_id: String,
    pub full_name: String,
    pub normalized_name: String,
    pub answers: Vec<i64>,
    pub user_agent: String,
    pub ip: String,
    pub location_id: Option<String>,
    pub geo_status: String,
}

#[derive(Debug)]
pub struct FinalizeSpinResult {
    pub finalized: bool,
    pub previous_prize: Option<String>,
}

impl SubmissionModel {
    pub async fn find_by_id(
        state: &AppState,
        paths: &CampaignPaths,
        id: &str,
    ) -> ApiResult<Option<Map<String, Value>>> {
        let parent = paths.parent_str(&state.db.client)?;
        state
            .db
            .client
            .fluent()
            .select()
            .by_id_in(SUBMISSIONS_SUBCOL)
            .parent(parent)
            .obj()
            .one(id)
            .await
            .map_err(|e| ApiError::Internal(e.into()))
    }

    pub async fn find_page(
        state: &AppState,
        paths: &CampaignPaths,
        limit: usize,
        _cursor: Option<Map<String, Value>>,
    ) -> ApiResult<(Vec<Map<String, Value>>, Option<Map<String, Value>>, bool)> {
        let parent = paths.parent_str(&state.db.client)?;
        let cap = limit.clamp(1, 200);
        let items: Vec<Map<String, Value>> = state
            .db
            .client
            .fluent()
            .select()
            .from(SUBMISSIONS_SUBCOL)
            .parent(parent)
            .order_by([
                ("submittedAt", FirestoreQueryDirection::Descending),
                ("__name__", FirestoreQueryDirection::Descending),
            ])
            .limit(cap as u32)
            .obj()
            .query()
            .await
            .map_err(|e| ApiError::Internal(e.into()))?;

        let has_more = items.len() == cap;
        Ok((items, None, has_more))
    }

    pub async fn ids_that_exist(
        state: &AppState,
        paths: &CampaignPaths,
        ids: &[String],
    ) -> ApiResult<HashSet<String>> {
        let mut existing = HashSet::new();
        for id in ids {
            if Self::find_by_id(state, paths, id).await?.is_some() {
                existing.insert(id.clone());
            }
        }
        Ok(existing)
    }

    pub async fn find_for_raffle_pool(
        state: &AppState,
        paths: &CampaignPaths,
        min_score: i64,
        prize_winners_only: bool,
    ) -> ApiResult<Vec<Map<String, Value>>> {
        let parent = paths.parent_str(&state.db.client)?;
        let rows: Vec<Map<String, Value>> = if min_score > 0 {
            state
                .db
                .client
                .fluent()
                .select()
                .from(SUBMISSIONS_SUBCOL)
                .parent(parent)
                .filter(|q| q.field("score").greater_than_or_equal(min_score))
                .obj::<Map<String, Value>>()
                .query()
                .await
                .map_err(|e| ApiError::Internal(e.into()))?
        } else {
            state
                .db
                .client
                .fluent()
                .select()
                .from(SUBMISSIONS_SUBCOL)
                .parent(parent)
                .obj::<Map<String, Value>>()
                .query()
                .await
                .map_err(|e| ApiError::Internal(e.into()))?
        };

        Ok(if prize_winners_only {
            rows.into_iter()
                .filter(|s| {
                    matches!(
                        s.get("prize").and_then(|v| v.as_str()),
                        Some(p) if p != "Nothing" && p != "pending"
                    )
                })
                .collect()
        } else {
            rows
        })
    }

    pub async fn create(
        state: &AppState,
        paths: &CampaignPaths,
        input: SubmissionCreateInput,
    ) -> ApiResult<Map<String, Value>> {
        let questions = QuestionModel::find_all(state, paths).await?;
        if questions.is_empty() {
            return Err(ApiError::WithStatus {
                status: axum::http::StatusCode::INTERNAL_SERVER_ERROR,
                message: "No questions configured.".into(),
                code: Some("NO_QUESTIONS".into()),
                data: None,
            });
        }

        if input.geo_status == "outside" {
            return Err(ApiError::with_code(
                axum::http::StatusCode::FORBIDDEN,
                "GEO_OUTSIDE_ZONE",
                "Participation is restricted to allowed geographic zones.",
            ));
        }

        if input.answers.len() != questions.len() {
            return Err(ApiError::WithStatus {
                status: axum::http::StatusCode::BAD_REQUEST,
                message: "answers must match the number of questions.".into(),
                code: Some("INVALID_ANSWERS_LENGTH".into()),
                data: None,
            });
        }

        let mut score = 0i64;
        for (i, q) in questions.iter().enumerate() {
            let ans = input.answers[i];
            let options_len = q
                .get("options")
                .and_then(|v| v.as_array())
                .map(|a| a.len())
                .unwrap_or(0) as i64;
            let correct = q.get("correctIndex").and_then(|v| v.as_i64()).unwrap_or(-1);
            if ans < 0 || ans >= options_len {
                return Err(ApiError::WithStatus {
                    status: axum::http::StatusCode::BAD_REQUEST,
                    message: format!("Invalid answer index for question {i}."),
                    code: Some("INVALID_ANSWER_INDEX".into()),
                    data: None,
                });
            }
            if ans == correct {
                score += 1;
            }
        }

        let total = questions.len() as i64;
        let percentage = ((score as f64 / total as f64) * 100.0).round() as i64;
        let prize = if score == total { "pending" } else { "Nothing" };
        let status = if prize == "pending" { "pending" } else { "completed" };

        let parent = paths.parent_str(&state.db.client)?;
        let db = state.db.client.clone();
        let session_id = input.session_id.clone();
        let payload = json!({
            "sessionId": input.session_id,
            "fullName": input.full_name.to_uppercase(),
            "normalizedName": input.normalized_name,
            "score": score,
            "total": total,
            "percentage": percentage,
            "prize": prize,
            "answers": input.answers,
            "userAgent": input.user_agent,
            "ip": input.ip,
            "locationId": input.location_id,
            "geoStatus": input.geo_status,
            "status": status,
            "submittedAt": millis_now(),
        });

        let session_id_for_tx = session_id.clone();
        let payload_for_tx = payload.clone();
        let prize_owned = prize.to_string();
        let status_owned = status.to_string();
        let parent_for_tx = parent.clone();
        let result = db
            .run_transaction(move |db, transaction| {
                create_tx(
                    db,
                    transaction,
                    parent_for_tx.clone(),
                    session_id_for_tx.clone(),
                    payload_for_tx.clone(),
                    score,
                    percentage,
                    &prize_owned,
                    &status_owned,
                )
            })
            .await
            .map_err(map_submission_error)?;

        Ok(result)
    }

    pub async fn delete(state: &AppState, paths: &CampaignPaths, id: &str) -> ApiResult<()> {
        let sub = Self::find_by_id(state, paths, id).await?;
        let reg = RegistrationModel::find_by_id(state, paths, id).await?;

        let mut decrement: Option<(String, String)> = None;
        if let Some(sub) = &sub {
            if let (Some(prize_name), Some(location_id)) = (
                sub.get("prize").and_then(|v| v.as_str()),
                sub.get("locationId").and_then(|v| v.as_str()),
            ) {
                if prize_name != "pending" && prize_name != "Nothing" {
                    let prizes = PrizeModel::find_all(state, paths).await?;
                    if let Some(prize) = prizes.iter().find(|p| {
                        p.get("name").and_then(|n| n.as_str()) == Some(prize_name)
                            && p.get("isRealPrize").and_then(|v| v.as_bool()).unwrap_or(true)
                    }) {
                        if let Some(prize_id) = prize.get("id").and_then(|v| v.as_str()) {
                            decrement = Some((location_id.to_string(), prize_id.to_string()));
                        }
                    }
                }
            }
        }

        let parent = paths.parent_str(&state.db.client)?;
        let writer = state.db.batch_writer().await.map_err(|e| ApiError::Internal(e.into()))?;
        let mut batch = writer.new_batch();

        delete_in_batch(&state.db.client, &mut batch, parent.as_str(), SUBMISSIONS_SUBCOL, id)?;
        delete_in_batch(&state.db.client, &mut batch, parent.as_str(), SESSIONS_SUBCOL, id)?;

        if reg.is_some() {
            delete_in_batch(
                &state.db.client,
                &mut batch,
                parent.as_str(),
                REGISTRATIONS_SUBCOL,
                id,
            )?;
            if let Some(reg) = reg {
                if let Some(normalized) = reg.get("normalizedName").and_then(|v| v.as_str()) {
                    delete_in_batch(
                        &state.db.client,
                        &mut batch,
                        parent.as_str(),
                        REGISTRATIONS_SUBCOL,
                        &format!("name_{normalized}"),
                    )?;
                }
            }
        }

        batch.write().await.map_err(|e| ApiError::Internal(e.into()))?;

        if let Some((location_id, prize_id)) = decrement {
            InventoryRepository::decrement_on_delete(state, paths, &location_id, &prize_id).await?;
        }

        Ok(())
    }
}

fn create_tx<'b>(
    db: firestore::FirestoreDb,
    transaction: &'b mut firestore::FirestoreTransaction,
    parent: String,
    session_id: String,
    payload: Value,
    score: i64,
    percentage: i64,
    prize: &str,
    status: &str,
) -> Pin<Box<dyn Future<Output = Result<Map<String, Value>, BackoffError<FirestoreError>>> + Send + 'b>> {
    let prize = prize.to_string();
    let status = status.to_string();
    Box::pin(async move {
        let session_doc: Option<Map<String, Value>> = db
            .fluent()
            .select()
            .by_id_in(SESSIONS_SUBCOL)
            .parent(parent.as_str())
            .obj()
            .one(&session_id)
            .await
            .map_err(BackoffError::Permanent)?;
        if session_doc.is_none() {
            return Err(BackoffError::Permanent(FirestoreError::InvalidParametersError(
                FirestoreInvalidParametersError::new(FirestoreInvalidParametersPublicDetails::new(
                    "NO_SESSION".to_string(),
                    "code".to_string(),
                )),
            )));
        }

        let sub_doc: Option<Map<String, Value>> = db
            .fluent()
            .select()
            .by_id_in(SUBMISSIONS_SUBCOL)
            .parent(parent.as_str())
            .obj()
            .one(&session_id)
            .await
            .map_err(BackoffError::Permanent)?;
        if let Some(existing) = sub_doc {
            let mut out = existing;
            out.insert("id".into(), json!(session_id));
            return Ok(out);
        }

        db.fluent()
            .update()
            .in_col(SUBMISSIONS_SUBCOL)
            .document_id(&session_id)
            .parent(parent.as_str())
            .object(&payload)
            .add_to_transaction(transaction)
            .map_err(BackoffError::Permanent)?;

        db.fluent()
            .update()
            .in_col(SESSIONS_SUBCOL)
            .document_id(&session_id)
            .parent(parent.as_str())
            .object(&json!({
                "hasPlayed": true,
                "playedAt": millis_now(),
                "score": score,
                "percentage": percentage,
                "prize": prize,
                "status": status,
            }))
            .add_to_transaction(transaction)
            .map_err(BackoffError::Permanent)?;

        let mut out = payload.as_object().cloned().unwrap_or_default();
        out.insert("id".into(), json!(session_id));
        Ok(out)
    })
}

fn delete_in_batch(
    db: &firestore::FirestoreDb,
    batch: &mut firestore::FirestoreBatch<'_, firestore::FirestoreSimpleBatchWriter>,
    parent: &str,
    collection: &str,
    id: &str,
) -> ApiResult<()> {
    db.fluent()
        .delete()
        .from(collection)
        .parent(parent)
        .document_id(id)
        .add_to_batch(batch)
        .map_err(|e| ApiError::Internal(e.into()))?;
    Ok(())
}

fn map_submission_error(err: FirestoreError) -> ApiError {
    let msg = err.to_string();
    if msg.contains("NO_SESSION") {
        return ApiError::WithStatus {
            status: axum::http::StatusCode::BAD_REQUEST,
            message: "No registration found for this session.".into(),
            code: Some("NO_SESSION".into()),
            data: None,
        };
    }
    ApiError::Internal(anyhow!(err))
}
